### ISSUE TITLE
Crossvalidation and error evaluation

### ISSUE THREAD

**Issue by catubc** [murthylab/sleap#580]
Hello

We have a quick question re: using the crossvalidation splitter built into Sleap 1.1.5.  We know how to run the code below to generate a trained NN and predictions

```
# Load data
labels = sleap.load_file(...)

# Generate a 0.8/0.1/0.1 split
labels_train, labels_val_test = labels.split(n=0.8)
labels_val, labels_test = labels_val_test.split(n=0.5)

# Save with images
labels_train.save("train.pkg.slp, with_images=True)
labels_val.save("val.pkg.slp, with_images=True)
labels_test.save("test.pkg.slp, with_images=True)
```

But we need two things:

1.  Are the manual labels already saved in the val.pkg.slp file?  We would need these to compare them to the prediction .slp file.  Ideally we would load them programmatically from the pkg.slp file.  Are they somewhere obvious?

2.  Also, is there a way to control how the split is done in the .split() method above?  This would be helpful if we wanted to control which frames were used - or wanted to do 10 x crossvalidation.

Any advice, especially re: 1. above, would be very helpful.

Thanks so much!

**Comment by talmo** [comment]
Hey Cat,

For 1, the answer is yes, they're saved in the labels file. If you `sleap.load_file("val.pkg.slp")`, you'll get a `sleap.Labels` object with all of the labeled frames for that split (as well as access to the raw image data).

This notebook might be informative: https://sleap.ai/notebooks/Model_evaluation.html

For 2, yeah that's a good point. I should have probably added a seed parameter to that function but I was trying to keep it simple. You can, however, roll your own splitting method pretty easily now. Check out the implementation for `Labels.split()`:
https://github.com/murthylab/sleap/blob/1b27526220136d7ec7f41496ca6f5fff1b97c6d9/sleap/io/dataset.py#L744-L783

Basically you could use sklearn's splitting or cross validation functions to sample the indices that you want to use for each split and then just do something like:

```py
n_labels = len(labels)
idx_train, idx_val, idx_test = split(np.arange(n_labels))  # where split is your desired splitting method 
labels_train = labels.extract(idx_train, copy=True)
labels_val = labels.extract(idx_val, copy=True)
labels_test = labels.extract(idx_test, copy=True)
```

Let me know if you have any questions!

Talmo

**Comment by catubc** [comment]
Hey

So re: 2, I think I can figure that out, I didn't realize there's an extract method that takes indexes as input.  It should be trivial then to generate datasets with test data that does not overlap so we can generate crossvalidation results.

Re; 1, it's still not clear how to identify the manual vs. predicted labels.  We have this .pkg.slp file that comes from 23 short videos.  We only human labeled 500 frames but the total count is 4434 frames which seems to include all the frames (perhaps the .pkg file was not exported correctly).  Can you advise if we did something wrong in mixing human+predicted labels in the .pkg file? Or is there another label on the frames that distinguishes between human vs. predicted?

Thanks so much,
Cat



```
# Load sleap pkg file containing labeled data and images
labels = sleap.load_file('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp')

print ("labels.count: ", labels.count)

for k in range(labels.get_labeled_frame_count()):
    print (k, labels.labels[k])

labels.count:  <bound method Sequence.count of Labels(labeled_frames=4434, videos=23, skeletons=1, tracks=4)>
0 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=1813, instances=0)
1 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=666, instances=4)
2 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2294, instances=4)
3 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=444, instances=4)
4 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=481, instances=4)
5 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=518, instances=4)
6 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=555, instances=4)
7 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=592, instances=4)
8 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=1480, instances=4)
9 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=1517, instances=4)
10 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=1702, instances=4)
11 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2035, instances=4)
12 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2072, instances=4)
13 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2109, instances=4)
14 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2146, instances=4)
15 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2590, instances=4)
16 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2627, instances=4)
17 LabeledFrame(video=HDF5Video('/media/cat/1TB/dan/cohort1/sleap/day001.pkg.slp'), frame_idx=2960, instances=4)

...
```

**Comment by catubc** [comment]
Ok, so we re-exported a .pkg.slp file that contained only labeled frames, but it seems when we extract frames we are still extracting frames that are suggestions/predicted. 

Is there something we're doing wrong here?  We used the GUI to export labeled frames only.

Thanks!

```
# Load sleap pkg file containing labeled data and images
labels = sleap.load_file('/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp')

print ("labels.count: ", labels.count)

for k in range(len(labels.videos)):
    #print (k, labels.labels[k])
    print (k, labels.videos[k])
 
print (labels.videos[1].shape)

print ('')

idx_train = np.arange(450)
labels_train = labels.extract(idx_train) 
print (labels_train)

labels.count:  <bound method Sequence.count of Labels(labeled_frames=2397, videos=12, skeletons=1, tracks=4)>
0 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(0, 1024, 1280, 3), backend=HDF5Video)
1 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(20, 1024, 1280, 3), backend=HDF5Video)
2 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(50, 1024, 1280, 3), backend=HDF5Video)
3 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(90, 1024, 1280, 3), backend=HDF5Video)
4 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(115, 1024, 1280, 3), backend=HDF5Video)
5 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(50, 1024, 1280, 3), backend=HDF5Video)
6 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(10, 1024, 1280, 3), backend=HDF5Video)
7 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(10, 1024, 1280, 3), backend=HDF5Video)
8 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(20, 1024, 1280, 3), backend=HDF5Video)
9 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(30, 1024, 1280, 3), backend=HDF5Video)
10 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(25, 1024, 1280, 3), backend=HDF5Video)
11 Video(filename=/media/cat/256GB/dan/cohort1/cohort1_day_only_labeled_frames.pkg.slp, shape=(10, 1024, 1280, 3), backend=HDF5Video)
(20, 1024, 1280, 3)

Labels(labeled_frames=450, videos=12, skeletons=1, tracks=4)

```

**Comment by catubc** [comment]
Ok, so the issue seems to be that there might be a bug in the way the GUI exports .pkg.slp files. The solution is to use another extract step to sub-select user_labeled_frame_inds. Those can then be saved and trained on etc.

I attach a screen grab provided by @talmo , see in particular line 16 of how to do this.

Thanks!

![loading_user_frames_only_and_saving](https://user-images.githubusercontent.com/4267452/133246283-bf1cbccf-cd67-4db6-ae70-2e6ac1357d60.png)
